File: calc.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <string>
#include <stack>
#include <cmath>

unsigned short parens(std::vector<std::string> s, unsigned short i) {
  std::stack<unsigned short> st;
  for (unsigned short ii=i; ii < s.size(); ii++) {
    if (s[ii] == "(") {
      st.push(0);
    }
    else if (s[ii] == ")") {
      st.pop();
      if (st.empty()) {
        return ii;
      }
    }
  }
  return 0;
}

long double calc(std::vector<std::string> s) {
  std::vector<std::string> so(s);
  for (unsigned short i = 0; i <= s.size(); i++) {
    if (s[i] == "("){
      unsigned short d = parens(s, i);
      so.erase(so.begin()+i, so.begin()+d+1);
      std::vector<std::string> sb(s.begin()+i+1, s.begin()+d);
      so.insert(so.begin()+i, std::to_string(calc(sb)));
      return calc(so);
    }
  }
  for (unsigned short i = 0; i <= s.size(); i++) {
    if (s[i] == "**") {
      so.erase(so.begin()+i-1, so.begin()+i+2);
      so.insert(so.begin()+i-1, std::to_string(std::pow(stod(s[i-1]), stod(s[i+1]))));
      return calc(so);
    }
    else if (s[i] == "//") {
        so.erase(so.begin()+i, so.begin()+i+1);
        so.insert(so.begin()+i, std::to_string(std::sqrt(stod(s[i+1]))));
        return calc(so);
    }
  }
  for (unsigned short i = 1; i <= s.size(); i++) {
    if (s[i] == "*") {
      so.erase(so.begin()+i-1, so.begin()+i+2);
      so.insert(so.begin()+i-1, std::to_string(stod(s[i-1]) * stod(s[i+1])));
      return calc(so);
    }
    else if (s[i] == "/") {
        so.erase(so.begin()+i-1, so.begin()+i+2);
        so.insert(so.begin()+i-1, std::to_string(stod(s[i-1]) / stod(s[i+1])));
        return calc(so);
    }
  }
  for (unsigned short i = 1; i <= s.size(); i++) {
    if (s[i] == "+") {
      so.erase(so.begin()+i-1, so.begin()+i+2);
      so.insert(so.begin()+i-1, std::to_string(stod(s[i-1]) + stod(s[i+1])));
      return calc(so);
    }
    else if (s[i] == "-") {
        so.erase(so.begin()+i-1, so.begin()+i+2);
        so.insert(so.begin()+i-1, std::to_string(stod(s[i-1]) - stod(s[i+1])));
        return calc(so);
    }
  }
  return stod(s[0]);
}

int main(int argc, char *argv[]) {
  std::vector<std::string> sv;
  const char del = ' ';
  std::string val;
  std::string exp;
  if (argc == 2) {
    exp = argv[1];
  }
  else {
    std::cout << "Give an expression: ";
    std::getline(std::cin, exp);
  }
  std::stringstream streamExp(exp);
  while (std::getline(streamExp, val, del)) {
    sv.push_back(val);
  }
  std::cout << calc(sv) << std::endl;
  return 0;
}

File: calc.py
#!/usr/bin/env python3.10

from sys import argv
from math import sqrt

def parens(s: tuple, i: int):
    openers = {"(": ")"}
    closers = {v: k for k, v in openers.items()}
    stack = []
    result = []
    for ii in range(i, len(s)):
        c = s[ii]
        if c in openers:
            stack.append([c, ii])
        elif c in closers:
            if not stack:
                raise ValueError(f"Tried to close brace without an open at position {i}.")
            pair, idx = stack.pop()
            if pair != closers[c]:
                raise ValueError(f"Mismatche brace at position {i}.")
            if idx == i:
                return ii
    if stack:
        raise ValueError(f"No closing brace at position {i}.")
    return result

def calc(s: tuple):
    so = list(s)
    for i in range(len(s)):
        match s[i]:
            case "(":
                d = parens(s, i)
                sb = []
                for x in range(i, d+1):
                    sb.append(so[i])
                    so.pop(i)
                sb.pop(0)
                sb.pop()
                so.insert(i, calc(sb))
                return calc(tuple(so))
    for i in range(len(s)-1):
        match s[i]:
            case "**":
                so.pop(i-1)
                so.pop(i-1)
                so.pop(i-1)
                so.insert(i-1, float(s[i-1]) ** float(s[i+1]))
                return calc(tuple(so))
            case "//":
                so.pop(i)
                so.pop(i)
                so.insert(i, sqrt(float(s[i+1])))
                return calc(tuple(so))
    for i in range(1, len(s)-1):
        match s[i]:
            case "*":
                so.pop(i-1)
                so.pop(i-1)
                so.pop(i-1)
                so.insert(i-1, float(s[i-1]) * float(s[i+1]))
                return calc(tuple(so))
            case "/":
                so.pop(i-1)
                so.pop(i-1)
                so.pop(i-1)
                so.insert(i-1, float(s[i-1]) / float(s[i+1]))
                return calc(tuple(so))
    for i in range(1, len(s)-1):
        match s[i]:
            case "+":
                so.pop(i-1)
                so.pop(i-1)
                so.pop(i-1)
                so.insert(i-1, float(s[i-1]) + float(s[i+1]))
                return calc(tuple(so))
            case "-":
                so.pop(i-1)
                so.pop(i-1)
                so.pop(i-1)
                so.insert(i-1, float(s[i-1]) - float(s[i+1]))
                return calc(tuple(so))
    return s[0]
if __name__ == "__main__":
    try:
        print(calc(argv[1].split(" ")))
    except IndexError:
        a = input("Give an expression: ")
        print(calc(a.split(" ")))
